### PR TITLE
google: Integrated the Google Signin.

### LINKED ISSUE #5153: Add third party authentication from google
**Is your feature request related to a problem? Please describe.**

We have already implemented third party auth using facebook. Now we should do the same for google.

**Additional context**

Parent: #5115

### PR DESCRIPTION
**Pull request by Hudda**
Fixes #5153

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Complete the OAuth registration feature by Google.

#### Changes proposed in this pull request:

-
-
-

### CODE REVIEW

File: app/api/auth.py [hunk 1]
```diff
@@ -129,9 +129,35 @@ def login_user(provider, auth_code):
...
+            'redirect_uri': request.args.get('redirect_uri'),
+            'client_secret': provider_class.get_client_secret(),
+            'code': auth_code + '/' + add_auth_code
         }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        access_token = requests.post('https://www.googleapis.com/oauth2/v4/token', data=payload, headers=headers).json()
+        id_token = access_token['id_token']
+        user_details = requests.get('https://oauth2.googleapis.com/tokeninfo?id_token=' + id_token).json()
+        if get_count(db.session.query(User).filter_by(google_id=user_details['sub'])) > 0:
+            user = db.session.query(User).filter_by(google_id=user_details['sub']).one()
```
Comment (iamareebjamal) [inline app/api/auth.py]: Redundant query, just fetch the first user and see if it is none. This is fetching the user twice. First, just to see the count and then again

File: app/api/auth.py [hunk 2]
```diff
@@ -129,9 +129,35 @@ def login_user(provider, auth_code):
...
+                jsonify(user_id=user.id, email=user.email, google_login_hash=user.google_login_hash), 200)
+
+        user = User()
+        user.google_id = user_details['sub']
+        user.google_login_hash = random.getrandbits(128)
+        user.password = ''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(8))
+        if user_details['email']:
+            user.email = user_details['email']
+
+        save_to_db(user)
```
Comment (iamareebjamal) [inline app/api/auth.py]: Most of the code is same for facebook. Please DRY

**Comment by Akhelesh** [comment]
@iamareebjamal I suggest we transition from using the requests library to requests-oauthlib. This is already a project dependency but rarely used. It will help keep the code consistent among the different API vendors  and will be easy for future contributors to catch up quick without much confusion. I have also created an issue for the same https://github.com/fossasia/open-event-frontend/issues/2113

**Comment by iamareebjamal** [comment]
Why not use Flask OAuthlib or Flask Dance?

**Comment by Hudda** [comment]
@iamareebjamal how should I proceed to merge migrations.

**Comment by iamareebjamal** [comment]
Just do it manually

**Comment by mrsaicharan1** [comment]
You can use `python manage.py db heads` to show your conflicitng heads. 
`python manage.py db merge <id of first head> <id of second head>` to merge both migrations into one.

**Comment by Hudda** [comment]
@iamareebjamal I made the required changes.

File: migrations/versions/6b2e5cfdd0f0_.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+"""empty message
```
Comment (iamareebjamal) [inline migrations/versions/6b2e5cfdd0f0_.py]: Delete the empty migration

File: app/api/auth.py [hunk 4]
```diff
@@ -128,10 +128,29 @@ def login_user(provider, auth_code):
 
     elif provider == 'google':
         provider_class = GoogleOAuth()
-        payload = {
+        data = {
```
Comment (iamareebjamal) [inline app/api/auth.py]: Use Flask Dance

File: migrations/versions/6b2e5cfdd0f0_.py (hunk 3 above)
Comment (Hudda) [reply migrations/versions/6b2e5cfdd0f0_.py]: @iamareebjamal then it will show error for multiple heads.

Comment (iamareebjamal) [reply migrations/versions/6b2e5cfdd0f0_.py]: Use the merge head commands

Comment (Hudda) [reply migrations/versions/6b2e5cfdd0f0_.py]: This files is created after I ran the merge head command.

**Comment by Hudda** [comment]
@iamareebjamal I have implemented the feature using flask-dance as you suggested.

File: app/api/auth.py [hunk 5]
```diff
@@ -24,6 +26,17 @@
 
 auth_routes = Blueprint('auth', __name__, url_prefix='/v1/auth')
 
+# Google
+try:
+    client_id = GoogleOAuth.get_client_id()
+    client_secret = GoogleOAuth.get_client_secret()
+except ProgrammingError:
+    client_id = 'google_client_id'
```
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/api/auth.py]: Remove the try catch

**Comment by Hudda** [comment]
@iamareebjamal I have made some changes. Please review.

**Comment by Hudda** [comment]
@iamareebjamal Please review the changes.

File: app/__init__.py [hunk 6]
```diff
@@ -115,8 +115,9 @@ def create_app():
 
     # development api
     with app.app_context():
+        db.create_all()
```
Comment (iamareebjamal) [inline app/__init__.py]: Remove

Comment (Hudda) [reply app/__init__.py]: @iamareebjamal I didn't find any other way to solve the problem. It would be great if you tell me how to solve this.

Comment (iamareebjamal) [reply app/__init__.py]: Refactor your code in such a way that it does not query the database at the time of application bootstrap process

**Comment by mariobehling** [comment]
Priority low: Delay

**Pull request closed without merging**